### PR TITLE
VM Exercise

### PR DESCRIPTION
**Pull request by paulghaddad**
In this solution, I implemented: 

* Basic functionality
* Stretch goals
* Basic error-handling
* Memory protection

### CODE REVIEW

File: computer-systems/introduction-prework/vm.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+			// protect instructions segment of memory
+			if destAddr < dataSegmentLower || destAddr > dataSegmentUpper {
+				return fmt.Errorf("Invalid memory address. Cannot write to address: %v", destAddr)
+			}
+
+			memory[destAddr] = srcVal
+		case Add:
+			srcReg1 := memory[pc+1]
+			srcReg2 := memory[pc+2]
+			registers[srcReg1] = registers[srcReg1] + registers[srcReg2]
```
Comment (Cheerazar) [inline computer-systems/introduction-prework/vm.go:68]: If you wanted to for the `Sub`, `Subi`, `Addi`, and this one you could use the `+=` or `-=` operators to be a bit more concise.

File: computer-systems/introduction-prework/vm.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+			increment := memory[pc+2]
+			registers[srcReg] = registers[srcReg] + increment
+		case Subi:
+			srcReg := memory[pc+1]
+			increment := memory[pc+2]
+			registers[srcReg] = registers[srcReg] - increment
+		case Jump:
+			amount := memory[pc+1]
+			registers[0] = amount
+			continue
```
Comment (Cheerazar) [inline computer-systems/introduction-prework/vm.go:84]: Nice! I didn't even think of using a `continue` here in my own code.

**Review by Cheerazar — COMMENTED** [review]
This looks great! Your code is clear and approachable 😄

File: computer-systems/introduction-prework/vm.go (hunk 2 above)
Comment (rchopra) [reply computer-systems/introduction-prework/vm.go:84]: +1 this is nice.

**Review by rchopra — COMMENTED** [review]
This is really nice and clear. I especially like that you named what the arguments actually are within each case.

**Pull request merged**